### PR TITLE
feat(onboarding): PR 2 — onboarding tools + system prompt + deterministic eval (JOV-2132)

### PR DESCRIPTION
**Pull request by itstimwhite**
## Summary

PR 2 of 4 in the unified Jovie AI chat front-door (JOV-2132). This PR adds the foundation: 6 new onboarding LLM tools, a Stanley-inspired system prompt, deterministic access-decision logic, and Zod-shaped metadata schemas.

**Status: WIP draft.** This is the foundation commit. The handler replacement (replacing PR 1's 501 stub with the real LLM dispatch) lands in a follow-up commit on this branch.

Stacked on top of PR 1 (#8509 — merged). Builds on the session cookie, Turnstile gate, rate limiters, and claim endpoint from that PR.

## What ships in this commit

**`lib/chat/prompts/onboarding.ts`** — Stanley-inspired system prompt
- Voice: confident, warm, sharp. NOT customer-service polite. NO emoji.
- One question per turn. Never two. Never a "first then" sequence.
- Spotify pick is the wow moment (lead with \`searchSpotifyArtist\`)
- Goals in order: feel-seen → mom-test → tangible profile → qualify → next step
- Handles meta-questions ("how much does it cost?") inline without breaking flow
- Objection handling with concrete rebuttals + \`recordInterviewSignal\` logging

**`lib/chat/tools/onboarding-signals.ts`** — Zod-shaped interview signal schema
- Typed payload for \`recordInterviewSignal\` → \`chat_conversations.metadata.interviewSignals\`
- Fields: releaseStage, audienceBand, currentTool, objection, freeNote
- Append-only metadata with versioning (\`metadataVersion: 1\`)
- \`collapseInterviewSignals\` helper for read-time aggregation

**`lib/chat/tools/onboarding-access-eval.ts`** — deterministic decision logic
- Pure server function (NOT an LLM tool — LLM cannot hallucinate the decision)
- Decision rules (top match wins):
  1. Spotify verified → \`instant_access\`
  2. Spotify followers >= 1,000 → \`instant_access\`
  3. Audience band >= 5k_to_50k → \`instant_access\`
  4. Audience 500_to_5k + active release → \`instant_access\`
  5. >= 3 turns without qualifying signal → \`waitlist\` (force decision)
  6. Otherwise → \`needs_more_info\` (LLM asks another question)
- 9 test cases covering every branch

**`lib/chat/tool-schemas.ts`** — 6 new tool schemas + \`ONBOARDING_TOOLS\` array
- \`searchSpotifyArtist\` — open artist picker popout
- \`confirmSpotifyArtist\` — lock the picked artist, pull enrichment
- \`checkHandle\` — handle availability via existing \`/api/handle/check\`
- \`recordInterviewSignal\` — silently log qualifying signal
- \`proposeNextStep\` — trigger deterministic eval, render decision card
- \`proposeCheckout\` — render checkout (route-handoff default, Embedded behind separate flag)
- \`proposeSocialLink\` reused from authenticated tools
- **Deleted from earlier plan:** \`revealProfilePreview\` — reveal state derives from tool history, not from an LLM command (single source of truth)

## What does NOT ship in this commit (next commits)

- [ ] Replace \`/api/chat\` onboarding mode 501 stub with executeChatTurn dispatch
- [ ] Statsig kill-switch flag \`onboarding_chat_v2\` wiring (default off)
- [ ] Per-tool deterministic UI escape hatches
- [ ] Integration test driving a full transcript end-to-end

## Test plan

- [x] \`pnpm --filter @jovie/web run typecheck\` — clean
- [x] \`pnpm biome check\` — clean
- [x] \`vitest lib/chat/tools/onboarding-access-eval.test.ts\` — 9/9 pass
- [ ] After dispatch lands: end-to-end test against /api/chat with mode='onboarding'
- [ ] /qa --exhaustive on \`/start\` once PR 3 lands the route

## Linear

JOV-2132 — Unified Jovie AI chat front-door (PR 2 of 4)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Rich anonymous onboarding chat with guided flow: artist picker, handle checks, interview signal capture, and checkout/waitlist proposals.
  * Onboarding-only toolset and deterministic eligibility evaluator for instant access / waitlist / needs-more-info.
  * Structured capture and consolidation of per-turn interview signals.

* **Tests**
  * New unit tests covering access-evaluation logic, onboarding handler flows, and edge cases.

* **Bug Fixes / Reliability**
  * Stricter message validation, improved streaming/error handling, client-disconnect handling, telemetry, and session cookie management.

[Review Change Stack](https://app.coderabbit.ai/change-stack/JovieInc/Jovie/pull/8551)

### CODE REVIEW

**Comment by itstimwhite** [comment]
## Review Summary
**Verdict**: Changes Requested

This is a substantial PR that adds the foundation for the onboarding chat flow. The architecture is well-thought-out, with clear separation between LLM-driven conversation and deterministic decision logic. However, there are several critical issues that need to be addressed before this can be merged.

### Critical Issues

- **apps/web/lib/chat/tools/onboarding-access-eval.ts:1-145** — The diff is truncated mid-function. The `evaluateAccessSignal` function body is missing. This is a blocking issue — we cannot review the core decision logic without seeing the complete implementation. The test file references `MAX_INTERVIEW_TURNS_BEFORE_FORCE` which isn't visible in the diff either.

- **apps/web/lib/chat/tools/onboarding-signals.ts** — Not shown in the diff. The `interviewSignalSchema` is imported in `tool-schemas.ts` but the actual schema definition isn't visible for review. This is critical since it defines the shape of data that flows through the entire onboarding system.

- **apps/web/lib/chat/tools/onboarding-tool-impls.ts** — Not shown in the diff. `createOnboardingTurnState` and `buildOnboardingTools` are imported in `onboarding-handler.ts` but their implementations aren't visible. These are the actual tool implementations that will be called by the LLM.

- **apps/web/lib/chat/run.ts:180-196** — The `mode` parameter defaults to `'app'` which is good for backward compatibility, but the `artistContext` is cast with `as ArtistContext` when mode is not onboarding. If a caller passes `mode: 'app'` with `artistContext: null`, this will throw a runtime error. The type signature says `ArtistContext | null` but the implementation assumes it's non-null for app mode. This should either:
  1. Throw a clear error if `artistContext` is null in app mode, or
  2. Make `artistContext` required when `mode: 'app'` via a discriminated union type.

- **apps/web/app/api/chat/onboarding-handler.ts:272-279** — The `validateOnboardingMessages` function checks `messages.length === 0` returns an error, but the schema allows `messages` to be optional. If `messages` is undefined, `rawMessages` will be an empty array `[]`, which will then fail validation with "messages array must be non-empty". This is inconsistent — either make messages required in the schema, or handle the undefined case differently.

- **apps/web/app/api/chat/onboarding-handler.ts:295** — `getEntitlements('free')` is called outside the try-catch block. If this throws (e.g., missing entitlements config), the error won't be caught by the Sentry handler below. This should be moved inside the try block or have its own error handling.

### Warnings

- **apps/web/lib/chat/tool-schemas.ts:73-130** — The `ONBOARDING_TOOLS` array includes `proposeSocialLink` which is reused from the authenticated tool set. However, the `proposeSocialLink` schema in the authenticated tools section (not shown in diff) may have different validation requirements for anonymous users vs authenticated users. Need to verify that the existing schema works correctly without a userId.

- **apps/web/lib/chat/prompts/onboarding.ts:1-88** — The system prompt mentions specific pricing ($39/mo for Pro, $149/mo for Max). This is a maintenance burden — if pricing changes, this prompt must be updated. Consider referencing a constant or environment variable for pricing values, or at minimum adding a comment noting where pricing is defined.

- **apps/web/app/api/chat/onboarding-handler.ts:143-170** — The Statsig gate check returns 503 with a generic message. Consider whether this should return a more specific error code that the frontend can handle gracefully (e.g., showing a "coming soon" message vs. a generic error).

- **apps/web/lib/chat/run.ts:180-196** — The `selectKnowledgeContextForTurn` is still called even in onboarding mode (line 200), but its result is only used in the `else` branch. This is a wasted async call. Consider wrapping it in the else branch or adding a guard.

### Suggestions

- **apps/web/app/api/chat/onboarding-handler.ts:272-279** — The `validateOnboardingMessages` function manually checks message structure. Consider extracting a shared validation utility that can be used by both the onboarding handler and the main chat handler to avoid drift.

- **apps/web/lib/chat/tools/onboarding-access-eval.ts** — Consider adding a `version` field to the decision output for future schema migrations, similar to the `metadataVersion` in `onboarding-signals.ts`.

- **apps/web/lib/chat/prompts/onboarding.ts** — The prompt mentions "Heads up — I'll remember this chat..." as the privacy disclosure. Consider whether this needs to be more explicit about data retention policies or GDPR compliance.

- **apps/web/lib/chat/tool-schemas.ts:73-130** — The `proposeCheckout` tool schema includes an optional `plan` field. Consider whether the LLM should be making plan recommendations at all, or if this should always be deferred to the user's choice on the checkout page.

- **apps/web/app/api/chat/onboarding-handler.ts:226-270** — The telemetry object wraps Sentry calls in try-catch blocks. Consider extracting this into a shared utility to avoid repetition across handlers.

### What Looks Good

- **Deterministic decision logic**: The separation between LLM-driven conversation and server-side access evaluation is excellent. The LLM cannot hallucinate access decisions, which is a critical security and correctness property.

- **System prompt quality**: The Stanley-inspired prompt is well-crafted with clear voice guidelines, hard rules, and concrete examples. The "one question per turn" rule and the handling of meta-questions/objections show careful thought about the conversation flow.

- **Test coverage**: 9 test cases covering every branch of the access evaluation logic, including edge cases like null follower counts and turn cap boundaries.

- **Type safety**: The use of Zod schemas for tool inputs and the `readonly` modifier on the `AccessDecisionInput` interface show good TypeScript practices.

- **Error handling**: The handler properly distinguishes between client disconnects (499) and internal errors (500), with Sentry instrumentation throughout.

- **Cookie management**: The manual `buildSessionCookieHeader` function for streaming responses is a pragmatic solution to the streaming cookie problem.

- **Progressive enhancement**: The PR description clearly documents what ships now vs. what comes in follow-up commits, making the review scope clear.

**Note**: This review is based on the truncated diff. The missing files (`onboarding-signals.ts`, `onboarding-tool-impls.ts`, and the complete `onboarding-access-eval.ts`) need to be reviewed before this can be approved.

**Review by itstimwhite — COMMENTED** [review]
Automated code review from Hermes PR Review Bot (deepseek/deepseek-v4-flash).

File: apps/web/app/api/chat/onboarding-handler.ts [hunk 1]
```diff
@@ -226,44 +272,179 @@ export async function tryHandleAnonymousOnboardingChat(
...
+          typeof text === 'string' &&
+          text.length > MAX_ONBOARDING_MESSAGE_LENGTH
+        ) {
+          return `messages[${i}] text part exceeds ${MAX_ONBOARDING_MESSAGE_LENGTH} chars`;
+        }
+      }
+    }
+  }
+  return null;
 }
```
[the remark this answers is not in the document]
Comment (itstimwhite) [reply apps/web/app/api/chat/onboarding-handler.ts:463]: Refactored. Extracted `validateMessageShape` and `validateTextPartLengths` helpers; `validateOnboardingMessages` is now a single-level loop calling each. SonarCloud cognitive complexity should drop comfortably under 15 at each layer.

File: apps/web/lib/chat/run.ts [hunk 2]
```diff
@@ -164,16 +180,23 @@ export async function executeChatTurn(
...
+  const systemPrompt =
+    mode === 'onboarding'
+      ? ONBOARDING_SYSTEM_PROMPT
+      : buildSystemPrompt(artistContext as ArtistContext, releases, {
+          aiCanUseTools: planLimits.booleans.aiCanUseTools,
+          aiDailyMessageLimit: planLimits.limits.aiDailyMessageLimit,
+          insightsEnabled,
+          knowledgeContext:
+            selectKnowledgeContextForTurn(uiMessages) || undefined,
+        });
```
[the remark this answers is not in the document]
Comment (itstimwhite) [reply apps/web/lib/chat/run.ts]: Fixed. Replaced the `as ArtistContext` cast with a runtime guard: if `mode='app'` and `artistContext` is null, throw 'artistContext is required when mode is app' before `buildSystemPrompt` is called. Fails fast at the entry instead of crashing inside the prompt builder.

File: apps/web/lib/chat/tools/onboarding-access-eval.test.ts [hunk 3]
```diff
@@ -0,0 +1,111 @@
+import { describe, expect, it } from 'vitest';
+import {
+  evaluateAccessSignal,
+  MAX_INTERVIEW_TURNS_BEFORE_FORCE,
+} from './onboarding-access-eval';
+
+const EMPTY_SIGNAL = {};
```
[the remark this answers is not in the document]
Comment (itstimwhite) [reply apps/web/lib/chat/tools/onboarding-access-eval.test.ts:7]: Nitpick acknowledged. EMPTY_SIGNAL is intentional in the access-eval test — it represents the 'no signal yet' state that proposeNextStep evaluates against. The Zod `.refine` check on the schema is for the LLM tool-call path; the eval function itself doesn't validate the signal shape, just reads fields. Leaving as-is.

File: apps/web/app/api/chat/onboarding-handler.ts [hunk 4]
```diff
@@ -226,44 +272,179 @@ export async function tryHandleAnonymousOnboardingChat(
...
+  const parts = [
+    `${ONBOARDING_SESSION_COOKIE_NAME}=${value}`,
+    'Path=/',
+    'HttpOnly',
+    'SameSite=Lax',
+    `Max-Age=${60 * 60 * 24 * 7}`,
+  ];
+  if (isSecureEnv()) parts.push('Secure');
+  return parts.join('; ');
+}
```
[the remark this answers is not in the document]
Comment (itstimwhite) [reply apps/web/app/api/chat/onboarding-handler.ts:408]: Nitpick acknowledged. Will extract a SESSION_COOKIE_MAX_AGE_SECONDS constant in a small follow-up once we also need it on the parser side. For PR 2 the 7-day TTL only lives in two places (mint here + the verify cookie spec) so the duplication cost is low.

File: apps/web/tests/unit/api/chat/onboarding-handler.test.ts [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+    );
+    const req = makeRequest({
+      mode: 'onboarding',
+      messages: [{ role: 'banana', parts: [] }],
+    });
+    const result = await tryHandleAnonymousOnboardingChat(req, 'req-4');
+    expect(result?.status).toBe(400);
+    const body = await result?.json();
+    expect(body.errorCode).toBe('INVALID_MESSAGES');
+  });
```
[the remark this answers is not in the document]
Comment (itstimwhite) [reply apps/web/tests/unit/api/chat/onboarding-handler.test.ts:162]: Added an explicit 'returns 400 when the messages array is empty' test (req-3b). 8/8 handler tests pass.
